### ISSUE TITLE
if select is initially hidden, doesn't work

### ISSUE THREAD

**Issue by iamnader** [harvesthq/chosen#92]
This is because when setting the width it will return 0.  Using this technique works: http://www.foliotek.com/devblog/getting-the-width-of-a-hidden-element-with-jquery-using-width/

The code should check if the elmt.is(:visible) and if so use the above technique.  I'm happy to fix in a fork.  @hat is the procedure?  Just patch the coffeescript?  How do you generate js and js min?

**Comment by Vinorcola** [comment]
The problem here is that if you change the size of the window, the select always appears the same size as there was initialized. I think a better solution would be to resize the select when shown or when window is resized, rather than to guess the width it should take at initialization when it is hidden.

**Comment by floydqwz** [comment]
Is there no way to dynamically change the size of the chosen object? Since I know when to change the size (resize event or onShow or similar), it could be a workaround. Is there an API function to change the size of the chosen object? Or reinitialize somehow?

**Comment by ve3** [comment]
I'm just thinking...

```html
<select id="my_select" class="chosen-selectbox">
    <option value="">Please select</option>
    <option value="1">1</option>
    <option value="2">2</option>
</select>
```

How if we get ``.css('display')`` of select box and apply it into Chosen container element on activating.
Get the id or class of selector (for example ``$('.chosen-selectbox').chosen();``) and apply those id or class into chosen container class on activating too. 
For example. The result will be... ``<div id="my_select_chosen" class="chosen-container chosen-container-single chosen-selectbox" style="display:inline-block;">``

This way, the Chosen will be generated without mess around with the css ``width`` and it should works fine. Developers can apply stylesheet width to ``.chosen-container.chosen-selectbox`` as they want.

**Comment by bbrooks** [comment]
:+1: for a fix. Ran into this today

**Comment by wndxlori** [comment]
:+1: Wasted *way* too much time tracking this down.

**Comment by prophe05** [comment]
Still having a problem with size in modals. Any plan to fix it in 4.0 release?

**Comment by Dmatafonov** [comment]
@danjjohnson your fix does not fix problem when first element is much smaller then others

![1](https://cloud.githubusercontent.com/assets/8863854/6482636/8fea2a7e-c270-11e4-9005-77bc890b0e21.png)

**Comment by Dmatafonov** [comment]
based on @danjjohnson that's what worked for me

hack

find line
```javascript
return "" + this.form_field.offsetWidth + "px";
```

replace with
```javascript
var calculatedWidth = this.form_field.offsetWidth;
          // Fix for zero width on hidden inputs.
          if (0 == calculatedWidth) {
              var clone = this.form_field_jq.clone();
              clone.css({"left": 10000, "position": "fixed"});
              clone.appendTo('body');
              calculatedWidth = clone.outerWidth();
              clone.remove();
          }
          return "" + calculatedWidth + "px";
```

**Comment by plmarcelo** [comment]
$('select').chosen({width: '100%'}); by @pfiller worked for me in a collapsable panel.

Thanks

**Comment by SurajShukla** [comment]
How can I get rid of the chosen-search class as I am not able to load the assets in my project directory?

**Comment by djabraham** [comment]
I had to add a timeout, when loading static selects from a template that had not been rendered yet. Nothing in the hierarchy had metrics when it attempted initialization. Wrapping it in a zero timeout at least bumped initialization to the end of the event queue..

    // Found this in chosen.jquery.js and patched it as follows
    AbstractChosen = (function () {
        function AbstractChosen(form_field, options) {
            if (!AbstractChosen.browser_is_supported()) {
                return;
            }

            var self = this;
            window.setTimeout(function () {
                self.form_field = form_field;
                self.options = options != null ? options : {};
                self.is_multiple = self.form_field.multiple;
                self.set_default_text();
                self.set_default_values();
                self.setup();
                self.set_up_html();
                self.register_observers();
            }, 0);
        }

**Comment by sobrinho** [comment]
I've just been bitten by this bug, 100% fixed the problem for me.

@harvesthq would you accept a pull request fixing this? Not sure about the best solution yet but I would be happy to fix and make a pull request :heart:

**Comment by erobwen** [comment]
A robust workaround for this is to use ng-if instead of ng-show. Works perfectly and is simple.  

I have figured out that ng-if is the prefered choice in many situations. Not only repairing broken ng-select, but also to allow recursive controllers (using a combination of ng-if, ng-init and ng-template).

**Comment by raffaeletani** [comment]
so it seems this 4 years old issue is still not resolved?!?!

**Comment by cederlof** [comment]
+1

**Comment by LauraLalune** [comment]
+1 in the meantime... I'm using select2

**Comment by Shakthivakeeswaran** [comment]
Is there any way to have a spinner while data is being appended to the chosen jquery. I am processing a huge number of records into chosen jquery and it takes time to process and show them. ??

**Comment by fsparv** [comment]
What an epic 5 year bug... and still biting people (ouch!).

**Comment by tanloverluke** [comment]
Google has a similar bug with their map that doesn't display properly in a hidden div. however both these problems have workarounds that are simple to implement.

**Comment by fsparv** [comment]
Yes, I hit that google maps bug some time last year IIRC, but how many folks spent half a day cursing and trying various things before finding this report? Bugs like these in popular waste hours of time. Multiplied across the number of folks learning either of these libraries each year the aggregate costs are probably staggering... If they aren't going to fix it, that's ok. Fixing it at this point might interact strangely with vast numbers of deployed workarounds perhaps... In that case they should admit they aren't fixing it and document the workaround somewhere prominent in the documentation so that folks don't struggle to make (or miss) deadlines over this stuff.

**Comment by brianjcohen** [comment]
My "overall" fix is similar to others, and it assumes that all my selects' widths in the markup are 100% of their parent div. We use bootstrap so this is the case for us.

`
$(".chosen-container").each(function(i, elem) {
	if ($(elem).width() == 0) {
		$(elem).width("100%");
	}
});
`

**Comment by geordi-la-forge** [comment]
To those of you that had a problem where if the element was initially inside a css hidden container, and then when made visible  had a width of 0px - the fix I found was to set the width in js config;

`    var config = {
  	      '.chosen-select'           : {},
  	      '.chosen-select-width'     : {width:"95%"}
  	    }
  	    for (var selector in config) {
  	      $(selector).chosen(config[selector]);
  	    }` 

Then I gave the select inputs the css class chosen-select-width 

`<select name="f01" class="chosen-select-width" id="item_select_staff">`

Seemed to do the trick with chosen v1.6.2 and jquery version 3.1.1.

**Comment by KZeni** [comment]
I just leave the default chosen width option (auto) and add
```
.chosen-container { min-width:5em; }
```
to my site's CSS (adjusting the min-width as needed). This makes it never become 0px (or just too small to use) while still fitting the select's ideal width when possible (forcing a 100% width or something to that effect doesn't work well when you have select elements inline with other content & this method addresses the core issue as well as allowing for inline select elements & seemingly offering the simplest & most straightforward implementation).

**Comment by bhartvigsen** [comment]
How is this still broken after SEVEN years?

**Comment by koenvanderlinden** [comment]
My workarond with bootstrap 3 is adding the following style:
```
.chosen-container {
    min-width: 90%;
}
```

**Comment by sharifanani** [comment]
> My workarond with bootstrap 3 is adding the following style:
> 
> ```
> .chosen-container {
>     min-width: 90%;
> }
> ```

Buy then the dropdown still doesn't show....

**Comment by qbadev** [comment]
it is already documented
https://harvesthq.github.io/chosen/options.html
Check the `width` option and its description

**Comment by KZeni** [comment]
@qbadev Although this makes it so the element doesn't size to fit the content when needed so I don't really see that as a fix (especially when using min-width provides a better result than this).

@koenvanderlinden recommends `min-width:90%;` so it's normally 90% width with it going up to 100% width if the contents of the select necessitate it. Then again, I still opt for `min-width:5em;` (exact em value varying depending on the use case) so it's never too small, but it still tries to expand to fit the content of the select rather than sizing based on the parent element.

**Comment by qbadev** [comment]
@KZeni I didn't say it's a fix. It's just documented. And it is a fine solution. My main issue was that Chosen was almost 0 width after init. Now, after "fix", I can see chosen and can do with it whatever I want using CSS classes applied to parent element.

I faced such problems many times before, that couldn't be calculated because object was 0 width. It can't be resolved in a universal way, as each application differs.
So when ones select is visible in a moment of init, no action is needed, but if it's hidden, using `width` option is a way to be used.

**Comment by odelgadillo** [comment]
Custom Width Support
Using a custom width with Chosen is as easy as passing an option when you create Chosen:
`$(".chosen-select").chosen({width: "95%"}); `


https://harvesthq.github.io/chosen/

**Comment by dmitryvhf** [comment]
I found this problem after jQuery upgrade from 1.* to 3.* release.
Solved that:

chosen.jquery.js:543
AbstractChosen.prototype.container_width = function () {
`return this.form_field.offsetWidth + "px";`
->
`return $(this.form_field).outerWidth() + "px";`
